### PR TITLE
[Enhance] ChoicesParam for LoudnessClient and PitchClient

### PR DESCRIPTION
**Pull request by jamesb93**
In keeping with the consistency of SpectralShape and BufStats, I think that Loudness should also include a ChoicesParam. This implements it in what I thought was the most similar way to SpectralShape and which is stylistically pretty much the same.

### CODE REVIEW

**Comment by tremblap** [comment]
I've added myself and Ted here as reviewers as this will need an interface discussion. To unpick:

- if we do it, pitch should have it too, obviously, as the only other non-multivariate descriptor
- in effect, neither of those actually need this interface per se if we stick to their original raison d'être: select was invented to remove an object in the workflow when one or more non-contiguous items were desired (mean and max for instance). The 'bonus' of providing the name of what channel is what was good but just because there was a lot of them... otherwise, we would need MFCC0 and MEL0 and all those guys, which are uselessly complicating the interface.
- I think that not having them in the 101 videos is more pedagogically fruitful than having them. @numchans is a bit harder but yields a lot more potential.

Now, we can discuss it all. if we agree on the last point, we can park this to v1 milestone.

**Comment by jamesb93** [comment]
> * if we do it, pitch should have it too, obviously, as the only other non-multivariate descriptor

I agree. The code is a little more entangled because the client handles scaling and replacing poor measurements (0hz) and I couldn't untangle it myself. It should, IMO, be a separated PR regardless.

> * in effect, neither of those actually need this interface per se if we stick to their original raison d'être: select was invented to remove an object in the workflow when one or more non-contiguous items were desired (mean and max for instance). The 'bonus' of providing the name of what channel is what was good but just because there was a lot of them... otherwise, we would need MFCC0 and MEL0 and all those guys, which are uselessly complicating the interface.

I think that this line of reasoning is needlessly overthinking it. People, including myself and those I have helped in their patching regularly just want one of the things out of these objects and this removes an object to do so and makes the code self documenting. It is unlikely you would decouple several MFCCs that are non-contiguous and separated from each other in a non-programmatic way. There is also the argument of consistency, in that these "semantic" multi-channel outputted descriptors (unlike mel/mfcc/stft) could be made more homogenous in their interface. This is good for all the reasons that are purported in previous discussions related to interface. I am thinking here of the differences between sourceA / source1 between objects, or references to interpolation being named differently.

> * I think that not having them in the 101 videos is more pedagogically fruitful than having them. @numchans is a bit harder but yields a lot more potential.

I've always found `numChans` and `startChan` hand-wavey to be honest. Again, it comes back to the notion of the code being expressive and self-documenting. We want loudness so we write loudness. We want confidence, we write confidence. Sometimes people will care which channel you need to extract from a multivariate descriptor, but in this case it's only ever 2 so enforcing people to care about the layout feels like an imposed difficulty that adds nothing to the usefulness of these objects. Furthermore, a `select` parameter doesn't exclude people using `fluid.bufselect~` or `numChans`. I would advocate any day to give people across a wide range of dispositions options. You like counting arrays and being particular about the layout of data and lots hate it and want to get The Thing :)

**Comment by tremblap** [comment]
I find most of your arguments compelling, although I am not certain that removing programmatic notions of the 101s is a good idea. We can talk about it quickly in tomorrow's meeting to see the best way forward with the various milestones and deadlines.

**Comment by tremblap** [comment]
Yes what you just say there are the arguments I read in the previous answer, and that I find very convincing. You know I love consistency of interface. we could even say I'm obsessed :)

File: include/clients/rt/LoudnessClient.hpp [hunk 1]
```diff
@@ -33,6 +34,7 @@ enum LoudnessParamIndex {
 };
 
 constexpr auto LoudnessParams = defineParameters(
+    ChoicesParam("select","Selection of Outputs","loudness","truepeak"),
```
Comment (weefuzzy) [inline include/clients/rt/LoudnessClient.hpp:37]: this isn't always true peak, so should probably just say peak

File: include/clients/rt/LoudnessClient.hpp [hunk 2]
```diff
@@ -122,6 +133,8 @@ class LoudnessClient : public FluidBaseClient, public AudioIn, public ControlOut
   algorithm::Loudness                                mAlgorithm;
   BufferedProcess                                    mBufferedProcess;
   FluidTensor<double, 1>                             mDescriptors;
+
+  index mMaxOutputSize;
```
Comment (weefuzzy) [inline include/clients/rt/LoudnessClient.hpp:137]: this is always 2, so no need for a variable

**Review by weefuzzy — CHANGES_REQUESTED** [review]

File: include/clients/rt/LoudnessClient.hpp (hunk 1 above)
Comment (jamesb93) [reply include/clients/rt/LoudnessClient.hpp:37]: Indeed - will change

File: include/clients/rt/LoudnessClient.hpp (hunk 2 above)
Comment (jamesb93) [reply include/clients/rt/LoudnessClient.hpp:137]: Okay cool! Less code is good :)

File: include/clients/rt/LoudnessClient.hpp (hunk 1 above)
Comment (jamesb93) [reply include/clients/rt/LoudnessClient.hpp]: Resolved in 3baac1d

File: include/clients/rt/LoudnessClient.hpp (hunk 2 above)
Comment (jamesb93) [reply include/clients/rt/LoudnessClient.hpp]: Resolved in b75b213

**Comment by jamesb93** [comment]
@weefuzzy can you please implement PitchClient thank you

**Review by weefuzzy — APPROVED** [review]

**Comment by weefuzzy** [comment]
This entails changes for SC classes as well : https://github.com/flucoma/flucoma-sc/pull/101

**Comment by tedmoore** [comment]
Should we just merge this in since the other `select` PRs have already been merged?

**Comment by jamesb93** [comment]
> Should we just merge this in since the other `select` PRs have already been merged?

I think so. I'm going to merge the core changes and then the SC classes so the nightlies work.

**Pull request merged**